bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler2/InternalSchedulerImpl.java

```diff
@@ -0,0 +1,298 @@
+ */
+@Component(name = "org.eclipse.smarthome.scheduler", service = InternalSchedulerImpl.class, immediate = true)
+public class InternalSchedulerImpl implements Scheduler {
+
+    private static final int THREAD_POOL_SIZE = 10;
+    private static final long THREAD_TIMEOUT = 65L;
+
+    private final List<Cron> crons = new ArrayList<>();
+    private final Logger logger = LoggerFactory.getLogger(InternalSchedulerImpl.class);
+
+    private final Clock clock = Clock.systemDefaultZone();
+    private ScheduledExecutorService executor;
+
+    @Activate
+    void activate() {
+        executor = Executors.newScheduledThreadPool(THREAD_POOL_SIZE);
+        ((ThreadPoolExecutor) executor).setKeepAliveTime(THREAD_TIMEOUT, TimeUnit.SECONDS);
+        ((ThreadPoolExecutor) executor).allowCoreThreadTimeOut(true);
+    }
+
+    @Deactivate
+    void deactivate() {
+        List<Runnable> shutdownNow = executor.shutdownNow();
+        if (shutdownNow != null && shutdownNow.size() > 0) {
+            logger.warn("Shutdown executables {}", shutdownNow);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Instant> after(Duration duration) {
+        CompletableFuture<Instant> deferred = new CompletableFuture<>();
+        Instant start = Instant.now();
+        ScheduledFuture<?> schedule = executor.schedule(() -> {
+            deferred.complete(start);
+        }, duration.toMillis(), TimeUnit.MILLISECONDS);
+        deferred.exceptionally((Throwable e) -> {
+            if (e instanceof CancellationException) {
+                schedule.cancel(true);
+            }
+            return null;
+        });
+        return deferred;
+    }
+
+    @Override
+    public <T> CompletableFuture<T> after(Callable<T> callable, Duration duration) {
+        CompletableFuture<T> deferred = new CompletableFuture<>();
+        ScheduledFuture<?> schedule = executor.schedule(() -> {
+            try {
+                deferred.complete(callable.call());
+            } catch (Throwable e) {
+                deferred.completeExceptionally(e);
+            }
+        }, duration.toMillis(), TimeUnit.MILLISECONDS);
+        deferred.exceptionally((Throwable e) -> {
+            if (e instanceof CancellationException) {
+                schedule.cancel(true);
+            }
+            return null;
+        });
+        return deferred;
+    }
+
+    static class Unique {
+        AtomicBoolean done = new AtomicBoolean();
+
+        interface RunnableException {
+            public void run();
+        }
+
+        boolean once(RunnableException o) {
+            if (done.getAndSet(true) == false) {
+                o.run();
+                return true;
+            } else {
+                return false;
+            }
+        }
+    }
+
+    @Override
+    public <T> CompletableFuture<T> before(CompletableFuture<T> promise, Duration timeout) {
+        CompletableFuture<T> d = new CompletableFuture<T>();
+        Unique only = new Unique();
```

Comment:
Would there really go something wrong without this `Unique`? It looks like that it should work without it.